devnet-sdk/testing/systest/multi_client.go

```diff
@@ -186,15 +186,54 @@ func (mc *MultiClient) fetchWithConsistencyCheck(
 		return nil, err
 	}
 
+	// Get the state root for the primary
+	primaryHeader, ok := primaryItem.(*types.Header)
+	if !ok {
+		// If we're fetching a block instead of a header
+		if block, ok := primaryItem.(*types.Block); ok {
+			primaryHeader = block.Header()
+		}
+	}
+
 	// Create a hash-only getter for followers
 	getFollowerHash := func(client HeaderProvider, num *big.Int) (common.Hash, error) {
-		_, _, hash, err := queryFn(client, num)
-		return hash, err
+		item, _, hash, err := queryFn(client, num)
+		if err != nil {
+			return common.Hash{}, err
+		}
+
+		// If hashes don't match, also check state roots
```

Comment:
We can almost assume it to be impossible for the hash collision to occur with not matching state roots

(one in 2^256 possibilities lol)